### PR TITLE
Fix: Pressing and releasing a key before the framework has started should not throw

### LINKED ISSUE #87391: Hot restart breaks keyboard input
Steps to reproduce:

1. Create an app with a TextField.
2. Launch the app with `flutter run -d linux`.
3. Type in the text field. This works fine.
4. Hot restart the app.
5. Type in the text field. This fails with:
```
** (fwml:2822048): WARNING **: 22:35:22.785: Unable to retrieve framework response: Message is not valid JSON

══╡ EXCEPTION CAUGHT BY SERVICES LIBRARY ╞══════════════════════════════════════════════════════════
The following assertion was thrown during a platform message callback:
A KeyUpEvent is dispatched, but the state shows that the physical key is not pressed. If this occurs
in real application, please report this bug to Flutter. If this occurs in unit tests, please ensure
that simulated events follow Flutter's event model as documented in `HardwareKeyboard`. This was the
event: KeyUpEvent#5dafa(physicalKey: PhysicalKeyboardKey#700e7(usbHidUsage: "0x000700e7", debugName:
"Meta Right"), logicalKey: LogicalKeyboardKey#0010e(keyId: "0x10000010e", keyLabel: "Super",
debugName: "Super"), character: null, timeStamp: 1115:12:48.298000)
'package:flutter/src/services/hardware_keyboard.dart':
Failed assertion: line 441 pos 16: '_pressedKeys.containsKey(event.physicalKey)'

Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md

When the exception was thrown, this was the stack:
#2      HardwareKeyboard._assertEventIsRegular.<anonymous closure> (package:flutter/src/services/hardware_keyboard.dart:441:16)
#3      HardwareKeyboard._assertEventIsRegular (package:flutter/src/services/hardware_keyboard.dart:452:6)
#4      HardwareKeyboard.handleKeyEvent (package:flutter/src/services/hardware_keyboard.dart:543:5)
#5      KeyEventManager.handleRawKeyMessage (package:flutter/src/services/hardware_keyboard.dart:821:35)
#6      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:73:49)
#7      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:72:47)
#8      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:379:35)
#9      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:376:46)
#10     _invoke2.<anonymous closure> (dart:ui/hooks.dart:203:15)
#14     _invoke2 (dart:ui/hooks.dart:202:10)
#15     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:42:5)
#16     _Channel.push (dart:ui/channel_buffers.dart:132:31)
#17     ChannelBuffers.push (dart:ui/channel_buffers.dart:329:17)
#18     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:536:22)
#19     _dispatchPlatformMessage (dart:ui/hooks.dart:90:31)
(elided 5 frames from class _AssertionError and dart:async)
════════════════════════════════════════════════════════════════════════════════════════════════════

### LINKED ISSUE #125975: Pressing and releasing a key before the framework has started throws
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Launch the code sample (tried on Linux and on Android).
2. Do a hot restart and very quickly focused the Flutter app, press and release a key (this might need several tries, because the key has to be pressed and released in the short time where the engine is ready to listen to platform keyboard events and the framework is not fully initialized).

### Expected results

No error

### Actual results

Error thrown: `flutter: A KeyUpEvent is dispatched, but the state shows that the physical key is not pressed.`

### Code sample

<details><summary>Code sample</summary>

```dart

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(
        body: Center(
          child: TextField(
          ),
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
flutter: ══╡ EXCEPTION CAUGHT BY SERVICES LIBRARY ╞══════════════════════════════════════════════════════════
flutter: The following assertion was thrown during a platform message callback:
flutter: A KeyUpEvent is dispatched, but the state shows that the physical key is not pressed. If this occurs
flutter: in real application, please report this bug to Flutter. If this occurs in unit tests, please ensure
flutter: that simulated events follow Flutter's event model as documented in `HardwareKeyboard`. This was the
flutter: event: KeyUpEvent#17c68(physicalKey: PhysicalKeyboardKey#a8a55(usbHidUsage: "0x00070014", debugName:
flutter: "Key Q"), logicalKey: LogicalKeyboardKey#12bb1(keyId: "0x00000061", keyLabel: "A", debugName: "Key
flutter: A"), character: null, timeStamp: 5:07:41.188000)
flutter: 'package:flutter/src/services/hardware_keyboard.dart':
flutter: Failed assertion: line 471 pos 16: '_pressedKeys.containsKey(event.physicalKey)'
flutter:
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new?template=2_bug.yml
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #2      HardwareKeyboard._assertEventIsRegular.<anonymous closure> (package:flutter/src/services/hardware_keyboard.dart:471:16)
flutter: #3      HardwareKeyboard._assertEventIsRegular (package:flutter/src/services/hardware_keyboard.dart:482:6)
flutter: #4      HardwareKeyboard.handleKeyEvent (package:flutter/src/services/hardware_keyboard.dart:601:5)
flutter: #5      KeyEventManager.handleRawKeyMessage (package:flutter/src/services/hardware_keyboard.dart:1021:37)
flutter: #6      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:212:49)
flutter: #7      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:481:35)
flutter: #8      _invoke2 (dart:ui/hooks.dart:174:13)
flutter: #9      _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:40:5)
flutter: #10     _Channel._drainStep (dart:ui/channel_buffers.dart:223:31)
flutter: (elided 4 frames from class _AssertionError and dart:async)
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════```

</details>


### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.10.0-3.0.pre.403, on Ubuntu 22.10
    5.19.0-41-generic, locale fr_FR.UTF-8)
    • Flutter version 3.10.0-3.0.pre.403 on channel master at
      /home/bruno/Nevercode/flutter
    • Upstream repository git@github.com:NevercodeHQ/flutter.git
    • FLUTTER_GIT_URL = git@github.com:NevercodeHQ/flutter.git
    • Framework revision 110fe75b5d (il y a 7 heures), 2023-05-03 02:11:35 -0400
    • Engine revision 0407a16a9b
    • Dart version 3.1.0 (build 3.1.0-66.0.dev)
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /home/bruno/Android/Sdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: /home/bruno/Produits/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build
      11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 15.0.7
    • cmake version 3.24.2
    • ninja version 1.11.0
    • pkg-config version 0.29.2

[✓] Android Studio (version 2021.3)
    • Android Studio at /home/bruno/Produits/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      11.0.13+0-b1751.21-8125866)

[✓] VS Code
    • VS Code at /snap/code/current
    • Flutter extension version 3.62.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.10 5.19.0-41-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 112.0.5615.49

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by dkwingsmt**
Fixes https://github.com/flutter/flutter/issues/125975, which fixes https://github.com/flutter/flutter/issues/87391.

### Background 

Flutter requires that the key event stream conforms "the key event rule". Simply speaking, a pointer repeat event or a pointer up event must occur only when key is considered held down according to Flutter's record, while a pointer down event must occur only when the key is considered released. In debug mode, Flutter throws an assertion on violation.

### Problem
The key event rule might be broken during the hot restart, because the event is transmitted and processed asynchronously through the message channel, and at the time the framework reinitializes the keyboard record, the engine might have sent events to the framework.

During hot restart, the Dart VM is recreated but the engine isn't. The engine retains the keyboard state from the last session, meaning there can be keys pressed, but the framework's state starts from zero. Therefore the framework must query the latest keyboard state (i.e. "what keys are being pressed right now?") before assigning the keyboard handler.

However, if any key events occur between when the channelBuffers is created and when getKeyboardState returns, these keys will have been sent out by the engine, unprocessed by any channel handlers and therefore buffered, but are also included in the returned engine state, causing the error.

For example:
When Dart VM hot restarts while key A is pressed, therefore engine.pressedKeys = { 'A' }
While framework is being initialized, key F on the physical keyboard is pressed.
Engine processes the physical event, records  engine.pressedKeys = {'A', 'F'} and sends out keyboard event PointerData(down, 'F')
There are currently no handlers for flutter/keyevent, therefore this key event is buffered.
The framework finally initializes the binding and _initKeyboard sends out getKeyboardState and gets the pressed state {'A', 'F'}.
The framework assigns handler to flutter/keyevent, and received PointerData(down, 'F'), and finds that F is already pressed according to its record, and asserts.

An earlier attempt https://github.com/flutter/flutter/pull/122885 tried to fix the issue by making the framework query the keyboard state right before it sets the key event handlers. This resolves most, but not all, of the issue, because the message queue can still buffer 1 event, which has been included in the init state sent by the engine and will also be sent to the event handler, causing a crash.

### Solution

This PR fixes this issue by clearing the method channels before assigning channel handlers. 

### Remaining question

Currently there are two problems I'm not sure:
1. One step of the test requires waiting until the callbacks are set (before sending the 3rd event, otherwise the 3rd event will cause the event buffers to pop the 2nd event). I can't think of an elegant way other than polling the status in an infinite loop. The poll shouldn't last very long, since the process should only be a `_keyboard.syncKeyboardState().then` and a `scheduleMicrotask`, but it's still ugly.
2. Resizing the channel buffers currently occur in the framework. It works, but according to `ChannelBuffer`'s doc it seems the resizing is designed to occur from the engine side. And also, since my intention is to clear the buffer (I don't care about buffer size afterwards though), should we add a `clearBuffer` method instead?
image

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Review by bleroux — COMMENTED** [review]
This PR looks very promising. I'm eager to see it merged because it is a great improvement for dev experience.
Just spotted two nits.

File: packages/flutter/test/services/hardware_keyboard_init_test.dart [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    );
+
+    ensureBindingIsNotInitialized();
+    WidgetsFlutterBinding.ensureInitialized();
+
+    // Wait until the binding starts listening to the event callbacks.
+    //
+    // There isn't an elegant way to listen to this, so this test has to poll.
+    while (true) {
+      await Future<void>.delayed(const Duration(milliseconds: 20));
```
Comment (Hixie) [inline packages/flutter/test/services/hardware_keyboard_init_test.dart:143]: we should definitely never write code like this :-)

doesn't the framework synchronously add its listeners? (if not, why not?)

**Comment by Hixie** [comment]
can't we just buffer all the events and not discard them? (by having the platform side configure the channel to not discard events, by sending it the message that increases the buffer size)

**Comment by Hixie** [comment]
(or alternatively, if we do want to just discard all messages, can't the platform side set the buffer size to zero?)

**Comment by dkwingsmt** [comment]
> can't we just buffer all the events and not discard them? (by having the platform side configure the channel to not discard events, by sending it the message that increases the buffer size)

I've thought about it as well. There are two reasons I don't think it's as good:
* Theoretically we don't know how many events should be buffered. For example, all keys can be pressed altogether right at the hot restart, and there will be over 100 events to be buffered. 
* In the end the framework just wants the keyboard state, instead of past events.

**Comment by dkwingsmt** [comment]
> (or alternatively, if we do want to just discard all messages, can't the platform side set the buffer size to zero?)

We can, but isn't this the same process? It's just moving the two lines from the framework to the engine.

**Comment by dkwingsmt** [comment]
There is indeed an alternative approach, that `KeyboardManager` is not rebuilt at all, so that the keyboard state never needs to be rebuilt.

**Comment by Hixie** [comment]
> > (or alternatively, if we do want to just discard all messages, can't the platform side set the buffer size to zero?)
> 
> We can, but isn't this the same process? It's just moving the two lines from the framework to the engine.

The difference is that if the engine does it, there's never anything buffered, so there's nothing for the framework to worry about. Conceptually it should always be the sending side sending these buffer size messages, not the receiving side.

**Comment by bleroux** [comment]
> The difference is that if the engine does it, there's never anything buffered, so there's nothing for the framework to worry about. Conceptually it should always be the sending side sending these buffer size messages, not the receiving side.

We tried doing that on the engine side, but there is a core issue. Here is a part of the summary I wrote in https://github.com/flutter/flutter/issues/125975#issuecomment-1835822981

"
Unfortunately, it turned out that this solution (setting channel buffer to 0) is not applicable to very core channels such as lifecycle and keyboard because those channels are created very early on the engine side (before the Dart isolate has started and this is a blocker because the channel resizing logic is done on the Dart side).
When experimenting with it, I reached the same issue that Greg described in https://github.com/flutter/flutter/issues/131452#issuecomment-1677962781.
"


And I think this also applies to 
> can't we just buffer all the events and not discard them? (by having the platform side configure the channel to not discard events, by sending it the message that increases the buffer size)

The engine can receive some key events before the Dart isolate has started and at that time the buffering logic is not available because it is written in Dart.  


> doesn't the framework synchronously add its listeners? (if not, why not?)

The framework add the listeners asynchronously because it is waiting for the result of a method channel invocation to get the initial pressed state, see https://github.com/flutter/flutter/pull/122885.

**Comment by Hixie** [comment]
Maybe we should move the channel buffers logic up to C++? It sounds like there's a race condition between when the messages can start getting sent, and when the engine side is ready to start buffering them.

> The framework add the listeners asynchronously because it is waiting for the result of a method channel invocation to get the initial pressed state, see #122885.

We probably should have some sort of callback or Future or binding API to indicate that the keyboard subsystem is ready. This all gets triggered by the binding, so maybe the test should override the binding and hook into the appropriate API to know when it's ready. (This may require some rejiggering of how the keyboard subsystem bootstraps, but that's ok. The key is to do it in a way that doesn't reserve memory in every Flutter app that's for no purpose other than serving this test, of course!)

**Comment by dkwingsmt** [comment]
Closing it for now since I don't have the bandwidth. Let me revisit it in the future.

**Pull request closed without merging**